pkg/metrics/logging_hook.go

```diff
@@ -0,0 +1,77 @@
+// Copyright 2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metrics
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/cilium/cilium/pkg/components"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+
+	"github.com/sirupsen/logrus"
+)
+
+// incErrorWarningMetric increments the error/warning Prometheus metric based on
+// the logger entry.
+func incErrorWarningMetric(entry *logrus.Entry) error {
+	// Every instance of Entry created from DefaultLogger is expected to
+	// have 'subsystem' field.
+	iSubsystem, ok := entry.Data[logfields.LogSubsys]
+	if !ok {
+		serializedEntry, err := entry.String()
+		if err != nil {
+			return fmt.Errorf("log entry cannot be serialized and doesn't contain 'subsys' field")
+		}
+		return fmt.Errorf("log entry doesn't comtain 'subsys' field: %s", serializedEntry)
+	}
+
+	subsystem, ok := iSubsystem.(string)
+	if !ok {
+		return fmt.Errorf("type of the 'subsystem' log entry field is not string but %s", reflect.TypeOf(iSubsystem))
+	}
+
+	ErrorsWarnings.WithLabelValues(entry.Level.String(), subsystem).Inc()
+
+	return nil
+}
+
+// LoggingHook is a hook for logrus which counts error, fatal, panic and warning
+// messages as a Prometheus metric.
+type LoggingHook struct{}
+
+// Levels returns the list of logging levels on which the hook is triggered.
+func (h *LoggingHook) Levels() []logrus.Level {
+	return []logrus.Level{
+		logrus.PanicLevel,
+		logrus.FatalLevel,
+		logrus.ErrorLevel,
+		logrus.WarnLevel,
+	}
+}
+
+// Fire is the main method which is called every time when logger has an error,
+// fatal, panic or warning message.
+func (h *LoggingHook) Fire(entry *logrus.Entry) error {
+	// NOTE(mrostecki): For now errors and warning metric exists only for Cilium
+	// daemon, but support of Prometheus metrics in some other components (i.e.
+	// cilium-health - GH-4268) is planned.
+	switch {
+	case components.IsCiliumAgent():
```

Comment:
This can be removed if you create a constructor for `LoggingHook` and you keep an attribute to know which component belongs to which `LoggingHook`.